query-engine/core/src/interactive_transactions/manager.rs

```diff
@@ -0,0 +1,201 @@
+
+pub static CLOSED_TX_CACHE_SIZE: Lazy<usize> = Lazy::new(|| match std::env::var("CLOSED_TX_CACHE_SIZE") {
+    Ok(size) => size.parse().unwrap_or(100),
+    Err(_) => 100,
+});
+
+pub struct ITXManager {
+    /// Stores all current transactions (some of them might be already committed/expired/rolled back).
+    ///
+    /// There are two tiers of locks here:
+    ///  1. Lock on the entire hashmap. This *must* be taken only for short periods of time - for
+    ///     example to insert/delete transaction or to clone transaction inside.
+    ///  2. Lock on the individual transactins. This one can be taken for prolonged periods of time - for
+    ///     example to perform an I/O operation.
+    ///
+    /// The rationale behind this design is to make shared path (lock on the entire hashmap) as free
+    /// from contention as possible. Individual transactions are not capable of concurrency, so
+    /// taking a lock on them to serialise operations is acceptable.
+    ///
+    /// Note that since we clone transaction from the shared hashmap to perform operations on it, it
+    /// is possible to end up in a situation where we cloned the transaction, but it was then
+    /// immediately removed by the background task from the common hashmap. In this case, either
+    /// our operation will be first or the background cleanup task will be first. Both cases are
+    /// an acceptable outcome.
+    transactions: Arc<RwLock<HashMap<TxId, Arc<RwLock<InteractiveTransaction>>>>>,
+
+    /// Cache of closed transactions. We keep the last N closed transactions in memory to
+    /// return better error messages if operations are performed on closed transactions.
+    closed_txs: Arc<RwLock<LruCache<TxId, ClosedTx>>>,
+
+    timeout_sender: UnboundedSender<TxId>,
+}
+
+impl ITXManager {
+    pub fn new() -> Self {
+        let transactions: Arc<RwLock<HashMap<TxId, Arc<RwLock<InteractiveTransaction>>>>> =
+            Arc::new(RwLock::new(HashMap::default()));
+        let closed_txs = Arc::new(RwLock::new(LruCache::new(*CLOSED_TX_CACHE_SIZE)));
+        let (timeout_sender, mut timeout_receiver) = unbounded_channel();
+
+        crosstarget_utils::task::spawn({
+            let transactions = transactions.clone();
+            let closed_txs = closed_txs.clone();
+            async move {
+                while let Some(tx_id) = timeout_receiver.recv().await {
+                    let transaction_entry = match transactions.write().await.remove(&tx_id) {
+                        Some(transaction_entry) => transaction_entry,
+                        None => {
+                            // Transaction was committed or rolled back already.
+                            continue;
+                        }
+                    };
+                    let mut transaction = transaction_entry.write().await;
+
+                    // If transaction was already committed, rollback will error.
+                    let _ = transaction.rollback(true).await;
```

Comment:
Could selectively ignore only "already committed" error here?

Reply:
There are 3 states transaction can be in here:
1. "Already committed" - we should not attempt a rollback
2. "Already rolled back" - again, we should not attempt the rollback, it was already done
3. "Open" - the only case where we can roll back

Why should we only ignore the first case and not the second one?

Reply:
Sorry, I mean to scope it to both "already commited" and "already rolled back".
By point was more for the case if open transaction fails to rollback for unexpected for us reason, so that we know it happened.